doc/adr/0001-jetstream-json-api-design.md

```diff
@@ -0,0 +1,134 @@
+# 1. JetStream JSON API Design
+
+Date: 2020-04-30
+
+## Status
+
+Proposed
+
+## Context
+
+At present, the API encoding consists of mixed text and JSON, we should improve consistency and error handling.
+
+## Decision
+
+We anticipate many types of consumer for JetStream including very small devices via MQTT and those written in 
+languages without easy to use JSON support like C.
+
+We will make it easy for light-weight clients by using `+OK` and `-ERR` style prefixes for client interactions,
+but the server management API will be JSON all the time. The rationale being that such small devices will most
+likely have their Consumers and Streams provisioned as part of the platform and often might not care for the
+additional context.
+
+In client responses we will use URL Encoding to provide additional context so multiple fields worth of
+context can be provided. URL Encoding has [broad language support](https://www.rosettacode.org/wiki/URL_decoding).
+
+### Admin APIs
+
+#### Requests
+
+All Admin APIs that today accept `nil` body should also accept an empty JSON document as request body.
+
+#### Responses
+
+All responses will be JSON objects, a few examples will describe it best. Any error that happens has to be 
+communicated within the originally expected message type. Even the case where JetStream is not enabled for
+an account, the response has to be the data type the API would normally generate with additional `error`.
```

Comment:
@ripienaar My reading of this is that it requires (as is shown below) fields that only make sense in the ok version to be present as well.
config / state make very little sense when there is an error.

I would want to adapt this wherever json is produced going forward.
So, when I look at this from the point of adopting it to the monitoring endpoints as system events and services I would have to add extra code to produce an empty config.

This would complicate more generic code.
Say you parse the input and it's not proper json, instead of returning with
type/time/error, now I also have to be able to produce empty objects...

@aricart you've done a lot more json than I, would this be allowable?

Reply:
This empty config that I showed above is literally a json.Marshal of `StreamInfo{}`, those default values is what is a `empty` version of StreamInfo - and also the minimum what the JSON Schemas will accept.  It's a horrible amount of work, but it's unfortunately right and Derek paved the way in JS code base to make really nice valid empty structs.

So if we have code in monitoring that doesnt do this I'd also like that it does that going forward.  And I'd also like JSON Schemas and `type`, `timestamp` and `id` as for the other advisories - I can show you that offline, but I'll demo this tomorrow in demo day thing.

Reply:
I hear you on the below comment. So you don't have data that becomes obsolete with the presence of an error. would changing line 36 as follows be ok and accommodate both?:
an account, the response has to be a valid data type with the addition of `error`.
When error is present empty fields may be omitted as long as the response still adheres to the schema. For the jetstream use case the schema does not allow for this.

Reply:
I'll add clarifying words - but it's implied, it has to be the original data type - if it's appropriate for that data type to be empty than thats fine.

Reply:
Updated but didnt add `For the jetstream use case the schema does not allow for this.`, because JetStream ones DO leave out empty fields where possible.